src/ert/_c_wrappers/util/substitution_list.py

```diff
@@ -1,141 +1,160 @@
 from __future__ import annotations
 
-import copy
+import logging
 import os
-from typing import Iterable, Tuple
+import re
+from typing import TYPE_CHECKING, Optional, Tuple
 
-from cwrap import BaseCClass
 from ecl.ecl_util import get_num_cpu as get_num_cpu_from_data_file
 
-from ert import _clib
-from ert._c_wrappers import ResPrototype
+logger = logging.getLogger(__name__)
+_PATTERN = re.compile("<[^<>]+>")
 
 
-class SubstitutionList(BaseCClass):
-    TYPE_NAME = "subst_list"
+# Python 3.8 does not implement UserDict as a MutableMapping, meaning it's not
+# possible to specify the key and value types.
+#
+# Instead, we only set the types during type checking
+if TYPE_CHECKING:
+    from collections import UserDict
 
-    _alloc = ResPrototype("void* subst_list_alloc()", bind=False)
-    _free = ResPrototype("void subst_list_free(subst_list)")
-    _size = ResPrototype("int subst_list_get_size(subst_list)")
-    _iget_key = ResPrototype("char* subst_list_iget_key(subst_list, int)")
-    _get_value = ResPrototype("char* subst_list_get_value(subst_list, char*)")
-    _has_key = ResPrototype("bool subst_list_has_key(subst_list, char*)")
-    _append_copy = ResPrototype("void subst_list_append_copy(subst_list, char*, char*)")
-    _alloc_filtered_string = ResPrototype(
-        "char* subst_list_alloc_filtered_string(subst_list, char*, char*, int)"
-    )
-    _deep_copy = ResPrototype("subst_list_obj subst_list_alloc_deep_copy(subst_list)")
+    _UserDict = UserDict[str, str]
+else:
+    from collections import UserDict as _UserDict
 
-    def __init__(self):
-        c_ptr = self._alloc(None)
 
-        if c_ptr:
-            super().__init__(c_ptr)
-        else:
-            raise ValueError("Failed to construct subst_list instance.")
-
-    @classmethod
-    def from_dict(cls, config_dict) -> SubstitutionList:
+class SubstitutionList(_UserDict):
```

Comment:
We should eventually just have `SubstitutionList` be a `Dict[str, str]` and have functions that work on that class, but we don't need to do that now.